### ISSUE TITLE
Teardown not really releasing connection

### ISSUE THREAD

**Issue by tom2strobl** [balderdashy/sails-postgresql#216]
We use the raw _waterline_ with the _postgresql adapter_ on Amazon Lambda, the problem is that everytime we initialize the db-connection per function it adds up until our database can't handle any more connections. `waterline.teardown()` to the rescue! Well not really, it looks like on https://github.com/balderdashy/sails-postgresql/blob/master/lib/adapter.js#L113 only the local connection object gets deleted, but the actual postgres connection is not closed. I'm not too familiar with the low-level API's, but shouldn't there be a `pg.end()` of some sort?
Or in other words: Is there a way to really release a connection without waiting for an automated timeout release? (Our poolSize is set to 1 to save on connection counts)

**Comment by devinivy** [comment]
Sounds like a bug.  Marking it as such!  Would love a failing test if possible.

**Comment by tom2strobl** [comment]
On the side a little more info as I'm investigating: I tried setting poolSize to 0 to disable pooling (actually don't need it for this case and thought it might auto-close connections then), but that didn't change anything about the outcome. The log (that will not help I suppose):
```
2015-12-17T14:46:32.146Z	Error creating a connection to Postgresql using the following settings: { host: '<hidden>', port: 5432, schema: true, ssl: false, user: '<hidden>', password: '<hidden>', database: '<hidden>', poolSize: 0, adapter: 'postgres', identity: '<hidden>' } 
2015-12-17T14:46:32.167Z	* * * Complete error details: { [error: sorry, too many clients already] name: 'error', length: 78, severity: 'FATAL', code: '53300', detail: undefined, hint: undefined, position: undefined, internalPosition: undefined, internalQuery: undefined, where: undefined, schema: undefined, table: undefined, column: undefined, dataType: undefined, constraint: undefined, file: 'proc.c', line: '339', routine: 'InitProcess' }
```

**Comment by tom2strobl** [comment]
On the `node-postgres` wiki it has this snippet:
```javascript
var pg = require('pg');
var connectionString = "pg://brian:1234@localhost/postgres"
pg.connect(connectionString, function(err, client, done) {
    client.query('SELECT name FROM users WHERE email = $1', ['brian@example.com'], function(err, result) {
      assert.equal('brianc', result.rows[0].name);
      done();  // client idles for 30 seconds before closing
    });
});
```
It says `client idles for 30 seconds before closing`, so this might be a `node-postgres` related "issue".

@devinivy Is there a public instance of the pg-object somewhere in the adapter so I could try to call `<something>.pg.end()` manually?

**Comment by particlebanana** [comment]
Ahh good catch @tom2strobl!. There are basically two options here:

1. Expose `pg.defaults` as config options.
2. Expose a flag that creates/closes a client per request using the PG [Client](https://github.com/brianc/node-postgres/wiki/Client#constructors) interface.

I think 1 is a bad idea because it sets these values globally on the `pg` module. This could lead to weird situations where you think you setting a per/connection setting.

The second option is perfect for this use case but is a lot of overhead for most use cases. If you would like to put together what you think an interface for this would look like that would be great.

**Comment by sailsbot** [comment]
Thanks for posting, @tom2strobl.  I'm a repo bot-- nice to meet you!

It has been 30 days since there have been any updates or new comments on this page.  If this issue has been resolved, feel free to disregard the rest of this message.  On the other hand, if you are still waiting on a patch, please:

  + review our [contribution guide](https://github.com/balderdashy/sails/blob/master/CONTRIBUTING.md) to make sure this submission meets our criteria (only _verified bugs_ with documented features, please;  no questions, commentary, or bug reports about undocumented features or unofficial plugins)
  + create a new issue with the latest information, including updated version details with error messages, failing tests, and a link back to [the original issue]().  This allows GitHub to automatically create a back-reference for future visitors arriving from search engines.

Thanks so much for your help!
